sidechain/test/src/sidechain_block_builder.rs

```diff
@@ -47,35 +48,46 @@ impl Default for SidechainBlockBuilder {
 	}
 }
 
-impl SidechainBlockBuilder {
-	pub fn random() -> Self {
+pub trait SidechainBlockBuilderTrait {
+	type Block: BlockT;
+	fn random() -> Self;
+	fn with_header(&mut self, header: Header) -> Self;
+	fn with_block_data(&mut self, block_data: BlockData) -> Self;
+	fn with_signer(&mut self, signer: ed25519::Pair) -> Self;
+	fn build(&self) -> Self::Block;
+	fn build_signed(&self) -> SignedBlock;
+}
+
+impl SidechainBlockBuilderTrait for SidechainBlockBuilder {
+	type Block = Block;
+	fn random() -> Self {
 		SidechainBlockBuilder {
 			signer: Pair::from_seed(&ENCLAVE_SEED),
 			header: SidechainHeaderBuilder::random().build(),
 			block_data: SidechainBlockDataBuilder::random().build(),
 		}
 	}
 
-	pub fn with_header(mut self, header: Header) -> Self {
+	fn with_header(&mut self, header: Header) -> Self {
 		self.header = header;
-		self
+		self.clone()
 	}
```

Comment:
Is there any reason for this? Builder patterns usually consume self, it is also more efficient. To prevent `mut self` we could also just make it `mut` internally.

```
fn with_header(self, header: Header) -> Self {
        let mut self_mut = self;
		self_mut.header = header;
		self_mut
	}
```